### PR TITLE
pip_audit: initial SBOM support via CycloneDX

### LINKED ISSUE #77: Emit results in a SBOM format
Blocked on https://github.com/trailofbits/pip-audit/issues/3.

### PR DESCRIPTION
**Pull request by woodruffw**
~~WIP; just pushing something up for visibility.~~

TODO:

- [x] Unit tests
- [x] Support for both XML and JSON SBOM outputs
- [x] Verify that vulnerability information is included in SBOMs

Relevant upstream tracking issues and PRs:

- [x] https://github.com/CycloneDX/cyclonedx-python-lib/issues/52
- [x] https://github.com/CycloneDX/cyclonedx-python-lib/issues/62
- [x] https://github.com/CycloneDX/cyclonedx-python-lib/pull/61

Closes #77.

### CODE REVIEW

File: pip_audit/format/cyclonedx.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+class PipAuditResultParser(BaseParser):
+    def __init__(self, result: Dict[service.Dependency, List[service.VulnerabilityResult]]):
+        super().__init__()
+
+        for (dep, vulns) in result.items():
+            c = Component(name=dep.name, version=str(dep.version))
+
+            # TODO(ww): Figure out if/how we want to include other dependency
+            # metadata in the BOM, such as author, license, etc.
```
Comment (woodruffw) [inline pip_audit/format/cyclonedx.py:22]: Braindump: we probably want to keep the `Dependency` model itself as small as possible, so I'm thinking something like this:

```python
@dataclass(frozen=True)
class Dependency:
    name: str
    version: Version
    metadata: Metadata = Metadata()
```

...where `Metadata` is a bunch of `Optional[T]`s that dependency sources can use to opportunistically enhance the SBOM output.

File: pip_audit/format/cyclonedx.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+            for vuln in vulns:
+                # TODO(ww): Figure out if/how we want to include other vulnerability
+                # metadata in the BOM, such as source (OSV/PyPI/etc), URL, etc.
+                c.add_vulnerability(
+                    Vulnerability(
+                        id=vuln.id,
+                        description=vuln.description,
+                        advisories=[f"Upgrade: {v}" for v in vuln.fix_versions],
```
Comment (woodruffw) [inline pip_audit/format/cyclonedx.py:31]: N.B.: The CycloneDX `Vulnerability` model doesn't have a field for upgrade advice, so I put it in a set of advisories instead.

File: pip_audit/format/cyclonedx.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+        for (dep, vulns) in result.items():
+            c = Component(name=dep.name, version=str(dep.version))
+
+            # TODO(ww): Figure out if/how we want to include other dependency
+            # metadata in the BOM, such as author, license, etc.
+
+            for vuln in vulns:
+                # TODO(ww): Figure out if/how we want to include other vulnerability
+                # metadata in the BOM, such as source (OSV/PyPI/etc), URL, etc.
```
Comment (woodruffw) [inline pip_audit/format/cyclonedx.py:26]: CycloneDX is smart enough generate a PyPI "PURL" identifier for us, so maybe this isn't necessary. But if we do want more metadata in each `VulnerabilityResult`, we could do something similar to the `Metadata` I'm thinking about above.

File: pip_audit/format/cyclonedx.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+            self._components.append(c)
+
+
+class CycloneDxFormat(VulnerabilityFormat):
+    def format(self, result: Dict[service.Dependency, List[service.VulnerabilityResult]]) -> str:
+        parser = PipAuditResultParser(result)
+        bom = Bom.from_parser(parser)
+
+        # TODO(ww): Configurable output format.
```
Comment (woodruffw) [inline pip_audit/format/cyclonedx.py:44]: CycloneDX supports both JSON and XML, the latter having various schemas:

https://github.com/CycloneDX/cyclonedx-python-lib/blob/main/cyclonedx/output/__init__.py

**Comment by woodruffw** [comment]
XML support is "in", but is currently failing with:

```
Traceback (most recent call last):
  File "/Users/william/devel/pip-audit/env/bin/pip-audit", line 33, in <module>
    sys.exit(load_entry_point('pip-audit', 'console_scripts', 'pip-audit')())
  File "/Users/william/devel/pip-audit/pip_audit/cli.py", line 215, in audit
    print(formatter.format(result))
  File "/Users/william/devel/pip-audit/pip_audit/format/cyclonedx.py", line 80, in format
    return formatter.output_as_string()  # type: ignore
  File "/Users/william/devel/pip-audit/env/lib/python3.6/site-packages/cyclonedx/output/xml.py", line 59, in output_as_string
    vulnerability=vulnerability))
  File "/Users/william/devel/pip-audit/env/lib/python3.6/site-packages/cyclonedx/output/xml.py", line 146, in _get_vulnerability_as_xml_element
    if vulnerability.has_ratings():
  File "/Users/william/devel/pip-audit/env/lib/python3.6/site-packages/cyclonedx/model/vulnerability.py", line 251, in has_ratings
    return len(self.get_ratings()) > 0
TypeError: object of type 'NoneType' has no len()
```

...so I'm guessing there's a part of the `Vulnerability` model that needs to be more initialized.

**Comment by woodruffw** [comment]
Upstream bug: https://github.com/CycloneDX/cyclonedx-python-lib/pull/61

**Comment by woodruffw** [comment]
Example XML output with an intentionally vulnerable requirements file:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<bom serialNumber="urn:uuid:5f03d520-4ca9-4025-8781-b0711cc98cda" version="1" xmlns="http://cyclonedx.org/schema/bom/1.3" xmlns:v="http://cyclonedx.org/schema/ext/vulnerability/1.0">
  <metadata>
    <timestamp>2021-11-09T18:46:43.718858+00:00</timestamp>
    <tools>
      <tool>
        <vendor>CycloneDX</vendor>
        <name>cyclonedx-python-lib</name>
        <version>0.10.0</version>
      </tool>
    </tools>
  </metadata>
  <components>
    <component bom-ref="pkg:pypi/docker-py@1.2.3" type="library">
      <name>docker-py</name>
      <version>1.2.3</version>
      <purl>pkg:pypi/docker-py@1.2.3</purl>
    </component>
    <component bom-ref="pkg:pypi/certifi@2021.10.8" type="library">
      <name>certifi</name>
      <version>2021.10.8</version>
      <purl>pkg:pypi/certifi@2021.10.8</purl>
    </component>
    <component bom-ref="pkg:pypi/charset-normalizer@2.0.7" type="library">
      <name>charset-normalizer</name>
      <version>2.0.7</version>
      <purl>pkg:pypi/charset-normalizer@2.0.7</purl>
    </component>
    <component bom-ref="pkg:pypi/idna@3.3" type="library">
      <name>idna</name>
      <version>3.3</version>
      <purl>pkg:pypi/idna@3.3</purl>
    </component>
    <component bom-ref="pkg:pypi/pip@21.3.1" type="library">
      <name>pip</name>
      <version>21.3.1</version>
      <purl>pkg:pypi/pip@21.3.1</purl>
    </component>
    <component bom-ref="pkg:pypi/requests@2.26.0" type="library">
      <name>requests</name>
      <version>2.26.0</version>
      <purl>pkg:pypi/requests@2.26.0</purl>
    </component>
    <component bom-ref="pkg:pypi/setuptools@40.6.2" type="library">
      <name>setuptools</name>
      <version>40.6.2</version>
      <purl>pkg:pypi/setuptools@40.6.2</purl>
    </component>
    <component bom-ref="pkg:pypi/six@1.16.0" type="library">
      <name>six</name>
      <version>1.16.0</version>
      <purl>pkg:pypi/six@1.16.0</purl>
    </component>
    <component bom-ref="pkg:pypi/urllib3@1.26.7" type="library">
      <name>urllib3</name>
      <version>1.26.7</version>
      <purl>pkg:pypi/urllib3@1.26.7</purl>
    </component>
    <component bom-ref="pkg:pypi/pyyaml@5.3" type="library">
      <name>pyyaml</name>
      <version>5.3</version>
      <purl>pkg:pypi/pyyaml@5.3</purl>
      <v:vulnerabilities>
        <v:vulnerability ref="pkg:pypi/pyyaml@5.3">
          <v:id>PYSEC-2020-96</v:id>
          <v:description>A vulnerability was discovered in the PyYAML library in versions before 5.3.1, where it is susceptible to arbitrary code execution when it processes untrusted YAML files through the full_load method or with the FullLoader loader. Applications that use the library to process untrusted input may be vulnerable to this flaw. An attacker could use this flaw to execute arbitrary code on the system by abusing the python/object/new constructor.</v:description>
          <v:recommendations>
            <v:recommendation>Upgrade</v:recommendation>
          </v:recommendations>
          <v:advisories>
            <v:advisory>Upgrade: 5.3.1</v:advisory>
          </v:advisories>
        </v:vulnerability>
        <v:vulnerability ref="pkg:pypi/pyyaml@5.3">
          <v:id>PYSEC-2021-142</v:id>
          <v:description>A vulnerability was discovered in the PyYAML library in versions before 5.4, where it is susceptible to arbitrary code execution when it processes untrusted YAML files through the full_load method or with the FullLoader loader. Applications that use the library to process untrusted input may be vulnerable to this flaw. This flaw allows an attacker to execute arbitrary code on the system by abusing the python/object/new constructor. This flaw is due to an incomplete fix for CVE-2020-1747.</v:description>
          <v:recommendations>
            <v:recommendation>Upgrade</v:recommendation>
          </v:recommendations>
          <v:advisories>
            <v:advisory>Upgrade: 5.4</v:advisory>
          </v:advisories>
        </v:vulnerability>
      </v:vulnerabilities>
    </component>
  </components>
</bom>
```

**Comment by woodruffw** [comment]
Turns out vulnerability information is intentionally not supported in the current JSON output format. Version 1.4 of CycloneDX will resolve that by promoting vulnerability information from an extension to a full-fledged part of the spec.

File: pip_audit/format/cyclonedx.py (hunk 1 above)
Comment (tetsuo-cpp) [reply pip_audit/format/cyclonedx.py:27]: I like that approach. 👍

**Review by tetsuo-cpp — APPROVED** [review]
Awesome! LGTM.

**Comment by woodruffw** [comment]
The only remaining blocker here is https://github.com/CycloneDX/cyclonedx-python-lib/pull/61. Once that's in, I'll merge this as an initial version and we can iterate on adding even more metadata.

File: pip_audit/format/cyclonedx.py (hunk 1 above)
Comment (woodruffw) [reply pip_audit/format/cyclonedx.py:27]: Yeah, I'll turn these TODOs into issues where we can figure out a reasonable approach to this.

Comment (woodruffw) [reply pip_audit/format/cyclonedx.py:27]: https://github.com/trailofbits/pip-audit/issues/123

**Pull request merged**